cmd/soroban-cli/src/config/secret.rs

```diff
@@ -140,6 +153,25 @@ impl Secret {
     }
 }
 
+pub enum StellarSigner {
+    Local(LocalKey),
+}
+
+#[async_trait::async_trait]
+impl Stellar for StellarSigner {
+    async fn get_public_key(&self) -> Result<stellar_strkey::ed25519::PublicKey, signer::Error> {
```

Comment:
Interestingly, generic `StellarSigner` can't provide public key, nor can it sign a blob. IMO the only thing it should do is just to sign a transaction envelope

Reply:
I'm confused by this comment. Again to maintain backwards compatibility with how we signed previously we signed the hash, which boils down to signing a blob or byte array. Also the `StellarSigner` trait uses default methods on the trait. So you only need to define these two methods and the rest will use these.

Reply:
An implementer of `Stellar` can implement all the methods, instead of relying on the default methods. And if it doesn't support `sign_blob` it can use `unreachable!()` inside the method. Because if it defines it's own version of `sign_txn_env`, `sign_blob` will never be called. The same applies for `get_public_key` if the implementation doesn't need it you don't have to implement it. However, it is used by `keys address ..`. This command retrieves the signer associated  and then resolves the public key if possible.  ledger and we would need to retrieve the key. This  would also require some work to allow this to work with lab as well.

Also which hardware wallets other than ledger support stellar and signing full transactions? Currently the ledger is in works to add that and only signs the hash... Either way I would assume that these wallets provide a way to access the associated public key. So for Lab, we could throw an error in the case of trying to do `keys address`, however, it would be nice to resolve the public key to use locally.

I do agree that the Stellar trait is too big and I did try with ways of inheriting the trait `trait Auth: Stellar`, but ran into some issues. And a lot of the design was guided initially by ledger and the initial implementation. And while it is easier to start with the more modular design, the refactor can be done later without affecting the end user's experience.

I personally don't know how to sign the full transaction with the local key. But it would be better than the hash.

Reply:
Generally, I agree. We can refactor signing methods later. As far as the public key method goes, I think it was only used to get source account, and as we discussed before it may no longer be needed.

Reply:
> As far as the public key method goes, I think it was only used to get source account

It's not though. It's used in `keys address` and generally in resolving a signer to a public key and will be used for auth signing.

Reply:
But should it be here, in generic Signer? It can be added separately in `keys address` and auth signing, while generic `--sign-with` doesn't require it

Reply:
Currently, we have Stellar trait with the following signature:
```rust
#[async_trait::async_trait]
pub trait Stellar {
    /// Currently only supports ed25519 keys
    async fn get_public_key(&self) -> Result<stellar_strkey::ed25519::PublicKey, Error>;

    /// Sign an abritatry byte array
    async fn sign_blob(&self, blob: &[u8]) -> Result<Vec<u8>, Error>;

    // Implementation functions below that use sign_blob (mainly)
    async fn sign_txn_hash(&self, txn: [u8; 32]) -> Result<DecoratedSignature, Error> { // impl }
    async fn sign_txn_env(
        &self,
        txn_env: TransactionEnvelope,
        network: &Network,
    ) -> Result<TransactionEnvelope, Error> { // impl}
    async fn sign_soroban_authorizations(
        &self,
        raw: &Transaction,
        network: &Network,
        expiration_ledger: u32,
    ) -> Result<Option<Transaction>, Error> { // impl }
    // Some extra wrappers over sign_tx_env and sign_soroban_authorizations 
}
```
    
Which I think is a bit too tailored for local signing. 
Generally, I don't think we need to be able to get a public key for signing just transactions (we only need source account, which is passed as an arg and is mandatory in all auto-signing requests). Just for signing we don't need a public key.
Same with signing a blob. We enforce blind signing for the wallet (which we should not do in a generic case, only in specific scenarios where we can't use clear signing).
Instead, I think we should only implement `sign_txn_env`

I understand that we also need to sign Soroban auth entries. For that we can just make a new trait. Not all signing methods might support this trait, and not all commands would need this trait (e.g. tx sign doesn't allow to sign only auth entries, same with most of commands. We only need to sign Soroban auth entries in `tx auth sign` and `tx invoke` -> which only work for the situation when we already have all local signers imported to begin with)

Proposed traits:
```rust
/// A trait for signing Stellar transactions
#[async_trait::async_trait]
pub trait Stellar {
    async fn sign_txn_env(
        &self,
        txn_env: TransactionEnvelope,
        network: &Network,
    ) -> Result<TransactionEnvelope, Error>
}

impl Stellar for LocalKey 
impl Stellar for Ledger
impl Stellar for Laboratory

/// A trait for signing Soroban authorization entries
#[async_trait::async_trait]
pub trait StellarAuth {
    async fn sign_txn_env(
        &self,
        txn_env: TransactionEnvelope,
        network: &Network,
    ) -> Result<TransactionEnvelope, Error> 
}

impl StellarAuth for LocalKey 
```    

WDYT? @willemneal @leighmcculloch

Reply:
I agree the interface is weak and reducing it to the single function would make the interface a better abstraction. Moving the auth into a separate interface too 👍🏻 . [The bigger the interface, the weaker the abstraction.](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=5m17s) For better or worse, it's common for Rust interfaces to look like this, similar to Java interfaces, with a mixin of utilities that make assumptions about the common case, but allow an "out" where you reimplement the utilities outside the common case. I think it'd be okay to merge with the large interface and then lets follow up with paring it down.

Reply:
I trimmed it down some, however, I kept it as signing a transaction since `sign_txn_env` can remain as a generic interface that will use a signer. After some thought I think that I had conflated two abstractions, a signer and a wallet. A wallet must have both a signer and a public address. 

I agree with @leighmcculloch that we could do some follow up work to address all this considering I am already working with adding the ledger, so we could meet up to discuss this in detail and how best to abstract this.

Reply:
I think the main issue is still there, `StellarSigner` must implement `Stellar` trait with `get_public_key` and `sign_blob` methods. Both of them are not required for envelope signing, only for auth signing. 
I don't mind revisiting abstractions later, but it looks like we may need to do it for both ledger and lab signing